src/storage-controller/src/lib.rs

```diff
@@ -2694,6 +2694,103 @@ where
 
         Ok(())
     }
+
+    async fn real_time_recent_timestamp(
+        &self,
+        source_ids: BTreeSet<GlobalId>,
+        timeout: Duration,
+    ) -> Result<
+        BoxFuture<'static, Result<Self::Timestamp, StorageError<Self::Timestamp>>>,
+        StorageError<Self::Timestamp>,
+    > {
+        use mz_storage_types::sources::GenericSourceConnection;
+
+        let mut rtr_futures = BTreeMap::new();
+
+        // Only user sources can be read from w/ RTR.
+        for id in source_ids.into_iter().filter(GlobalId::is_user) {
+            let collection = match self.collection(id) {
+                Ok(c) => c,
+                // Not a storage item, which we accept.
+                Err(_) => continue,
+            };
+
+            let (source_conn, remap_id) = match &collection.description.data_source {
+                DataSource::Ingestion(IngestionDescription {
+                    desc: SourceDesc { connection, .. },
+                    remap_collection_id,
+                    ..
+                }) => match connection {
+                    GenericSourceConnection::Kafka(_)
+                    | GenericSourceConnection::Postgres(_)
+                    | GenericSourceConnection::MySql(_) => {
+                        (connection.clone(), *remap_collection_id)
+                    }
+
+                    // These internal sources will never support RTR.
```

Comment:
Never is too strong a word; maybe they should support RTR?

Reply:
And lo, it came to pass...